### DISCUSSION TITLE
[Feature] Avoid duplicating assets and save server resources and storage space

### DISCUSSION THREAD

**Discussion by dpantel** [immich-app/immich#4542] [Feature Request]
### The feature

Media assets should be stored on a per-instance basis, rather than a per-user basis. This should be true for both uploaded media, as well as generated media.

Let's say I take a 20-minute video with my phone. Then I share it with my family members, who save it to their phones, and who all have the Immich app installed. Now Immich will upload multiple copies of the same video to the server, and then re-encode this video multiple times, once for each user.

Let's also say I have an External Library directory with 5000 photos of a family trip. I want to add that directory to each Immich account, so everyone can see those photos. Immich will now generate and store 5000 thumbnails for each user.

Basically, there is a lot of resource waste, both in storage, and in generating encoded video, thumbnails, etc. I think a lot can be saved by checking if an asset (uploaded or generated) already exists somewhere on the server before uploading/generating.

I think this can be accomplished relatively easily, either by co-mingling storage into one single pool for all users, or keeping assets nominally separate, but using hard-linking.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by v411e** [comment]
I would highly appreciate this feature. My use case is like this:
- I want to give my family members access to certain external libraries as we (obviously) share some common history like family vacations etc.
- As you pointed out, having all those Previews generated and stored multiple times on a per-user basis does waste a lot of resources and makes the whole multi-user thing a bit useless. I could also just create a dedicated Immich Instance per user and have a similiar resource footprint (at least storage wise)

**Comment by TheMarf** [comment]
This would be a great feature.  I had just commented on the immich-go repo asking about reducing duplicates because my wife and I have shared so many pictures with each other and there are so many duplicates of the same image even though they are attached to different user keys.  It sounds like this is likely something that needs to be addressed on the immich side.

**Comment by blublub** [comment]
I was looking to achieve that via a zfs partition in my Immich VM. Problem is that the copy in write feature of ZFS ruins that.
I have 1.8TB of image data and when I copy that onto a ZFS partition that vhdx has the size of 1.9TB and zpool shows a usage of 1TB - so 0.8TB could have been saved but got wasted in the process - so not worth it....

**Comment by Crocmagnon** [comment]
I'm looking for a photo backup solution that has this kind of feature, specifically with the iCloud Shared Photo Library usecase in mind. I'm currently uploading everything to nextcloud, but it doesn't manage duplicates either.

**Comment by DragonLingLuo** [comment]
I also would really appreciate this feature, and I think using hard links is a more user-friendly approach than mixing in a single pool. This makes it easy to back up manually while preserving the album structure without the need for an additional backup database (which stores which photos in a single pool correspond to which user's album).

**Comment by rosbeef** [comment]
I view this discussion just now as its not part of issue section.
I thinks it is a required functionality to save space, respect the planet and reduce server administration( increase disk storage )

I would ❤️ have this feature. Is their a branch with that somewhere ?

**Comment by rosbeef** [comment]
I have multiples ideas which i can not decide which one is the best:
- 1st : user have to Partner sharing to deduplicate photos
-- problem this case is not good for me as someone would delete a photo and the other not.
-- what happend if the owner want to delete the photo and the Partner not ?

- 2st : no matter the photos are shared between users or not, a process will deduplicate photos searching inside all users and add sharing before removing duplicated files.
-- problem which one will be the owner of the original photo.
-- the owner should be an immich root user and only give access to real users by sharing.
-- photos which are not duplicated stay inside the user folder.
-- how to recover photos for all users in case of database corruption

- 3nd : all photos are in immich root user folder and all photos are shared to real user depending on owner viewer rights. 
-- problem folder view option will show nothing
-- how to recover photos for all users in case of database corruption

- 4th : tell me

**Comment by easly1989** [comment]
Is there any news?
I think the best way to handle this would be to store all the photos in a specific (and unique) location. However, this would require Partner-level sharing settings, meaning the user would be informed that all photos are treated as if they were uploaded by a single user, with all the related benefits and drawbacks.

In this scenario, it’s clear that if one of the partners deletes a photo, it will be deleted for both.

A possible workaround could be to introduce a photo-locking mechanism: if one of the partners places a lock on a specific photo, that photo cannot be deleted by either of them unless the lock is removed; and it must be removed by the user who applied it.

What do you think?

**Reply by sdf121234** [reply]
Yes, I had the same idea as well. However, that wouldn’t be a solution for me personally: User A should not be able to see the photos of User B, but both should be able to see the photos of User C.

I’m currently testing the integration as an external library. While this doesn’t solve the issue with face tagging, it would at least solve access and viewing for me personally—hopefully. 😉

**Comment by rosbeef** [comment]
The 4 th idea could be to implement an s3 storage directly from immich without passing through posix fs. the s3 storage will manage duplicated blocks

**Reply by rosbeef** [reply]
#23745 maybe linked topic

**Comment by rosbeef** [comment]
Please do not close as this https://github.com/immich-app/immich/discussions/17422 . the 17422 was exactly se same and was closed for a wrong reason.
It is so important to save space in those times, even futo immich server should be aware of that because of the electronic price. And even if a family buy the online storage it should be so much space earned for more clients at less cost.
Please walk for the planet which is the base for future of immich ;)

**Comment by rosbeef** [comment]
@bo0tzz @mmomjian Sorry this is an open discussion of the same subject as https://github.com/immich-app/immich/discussions/17422
